services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,376 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// Packaging the return depending on if they want PCDs
+		n := int(math.Min(float64(len(outClusters)), float64(len(boxes)))) // should be same len but for safety
+		toReturn := make([]*vision.Object, n)
+		for i := 0; i < n; i++ { // for each cluster/box make an object
+			boxes[i].SetLabel("obstacle")
+			if o.returnPCDs {
+				pcdToReturn := pointcloud.New()
+				basicData := pointcloud.NewBasicData()
+				for _, pt := range outClusters[i].Observations {
+					if len(pt.Coordinates()) >= 3 {
+						vec := r3.Vector{X: pt.Coordinates()[0], Y: pt.Coordinates()[1], Z: pt.Coordinates()[2]}
+						err = pcdToReturn.Set(vec, basicData)
+						if err != nil {
+							return nil, err
+						}
+					}
+				}
+				toReturn[i] = &vision.Object{PointCloud: pcdToReturn, Geometry: boxes[i]}
+			} else {
+				toReturn[i] = &vision.Object{Geometry: boxes[i]}
+			}
+		}
+		return toReturn, nil
+	}
+}
+
+// makePointList will grab points from the depthmap. Sample every n.
+func (o *obsDepth) makePointList(n int) {
+	w, h := o.dm.Width(), o.dm.Height()
+	out := make([]obsPoint, 0, w*h)
+	for i := 0; i < w; i += n {
+		for j := 0; j < h; j++ {
+			out = append(out, obsPoint{point: image.Pt(i, j), depth: o.dm.GetDepth(i, j)})
+		}
+	}
+	o.ptList = out
+	o.ptChunks = splitPtList(out, chunkSize)
+}
+
+// splitPtList will split the pointlist into a list of lists, len(chunk) = chunkSize (not the last one).
+func splitPtList(slice []obsPoint, chunkSize int) [][]obsPoint {
+	var chunks [][]obsPoint
+	for i := 0; i < len(slice); i += chunkSize {
+		end := i + chunkSize
+
+		// necessary check to avoid slicing beyond
+		// slice capacity
+		if end > len(slice) {
+			end = len(slice)
+		}
+
+		chunks = append(chunks, slice[i:end])
+	}
+
+	return chunks
+}
+
+// isObstaclePoint returns true/false if compatible with another point in the depthmap.
+// as defined by Manduchi et al.
+func (o *obsDepth) isObstaclePoint(point image.Point) bool {
+	for _, p := range o.ptList {
+		if point == p.point {
+			continue
+		}
+		if o.isCompatible(point, p.point) {
+			return true
+		}
+	}
+	return false
+}
+
+// isCompatible will check compatibility between 2 points.
+// as defined by Manduchi et al.
+func (o *obsDepth) isCompatible(p1, p2 image.Point) bool {
+	// thetaMax in radians
+	xdist, ydist := math.Abs(float64(p1.X-p2.X)), math.Abs(float64(p1.Y-p2.Y))
+	zdist := math.Abs(float64(o.dm.Get(p1)) - float64(o.dm.Get(p2)))
+	dist := math.Sqrt((xdist * xdist) + (ydist * ydist) + (zdist * zdist))
+
+	if ydist < o.hMin || ydist > o.hMax {
+		return false
+	}
+	if ydist/dist < o.sinTheta {
+		return false
+	}
+	return true
+}
+
+// clustersToBoxes will turn the clusters we get from kmeans into boxes.
+func (o *obsDepth) clustersToBoxes(clusters clusters.Clusters) ([]spatialmath.Geometry, error) {
+	boxes := make([]spatialmath.Geometry, 0, len(clusters))
+
+	for i, c := range clusters {
+		var xmax, ymax, zmax float64
```

Comment:
Should these be something like negative infinity? In case all points have negative values?